File: Binary_Search/Search_a_2D_Matrix/solution.cpp
class Solution {
    public:
        bool searchMatrix(vector<vector<int>>& matrix, int target) {
            int n = matrix.size() ; 
            int m = matrix[0].size() ; 
            int l = 0 ; 
            int r = n*m - 1 ; 
            while( l < r ){
                int mid = l + ( r - l )/2 ; 
                int row = mid/m ; 
                int col = mid%m ; 
                if( matrix[row][col] > target ){
                    r = mid-1 ; 
                }else if( matrix[row][col] < target ){
                    l = mid + 1 ; 
                }else {
                    return true ; 
                }
            } 
            return matrix[ l/m ][ l%m ] == target ; 
        }
    };

File: Binary_Search/Search_a_2D_Matrix/solution.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        n = len(matrix) 
        m = len(matrix[0])
        l = 0  
        r = n*m - 1 
        while l < r : 
            mid = l + ( r - l )//2 
            row = mid // m 
            col = mid % m 
            if matrix[row][col] > target : 
                r = mid - 1 
            elif matrix[row][col] < target :  
                l = mid + 1 
            else : 
                return True 

        return matrix[l//m][l%m] == target